pkg/controller/rollout/rollout_bluegreen.go

```diff
@@ -0,0 +1,487 @@
+		tr := newTrafficRoutingContext(c)
+		done, err := m.trafficRoutingManager.DoTrafficRouting(tr)
+		blueGreenStatus.LastUpdateTime = tr.LastUpdateTime
+		if err != nil {
+			return err
+		} else if done {
+			blueGreenStatus.LastUpdateTime = &metav1.Time{Time: time.Now()}
+			blueGreenStatus.CurrentStepState = v1beta1.CanaryStepStateMetricsAnalysis
+			klog.Infof("rollout(%s/%s) step(%d) state from(%s) -> to(%s)", c.Rollout.Namespace, c.Rollout.Name,
+				blueGreenStatus.CurrentStepIndex, v1beta1.CanaryStepStateTrafficRouting, blueGreenStatus.CurrentStepState)
+		}
+		expectedTime := time.Now().Add(time.Duration(defaultGracePeriodSeconds) * time.Second)
+		c.RecheckTime = &expectedTime
+
+	case v1beta1.CanaryStepStateMetricsAnalysis:
+		klog.Infof("rollout(%s/%s) run bluegreen strategy, and state(%s)", c.Rollout.Namespace, c.Rollout.Name, v1beta1.CanaryStepStateMetricsAnalysis)
+		done, err := m.doCanaryMetricsAnalysis(c)
+		if err != nil {
+			return err
+		} else if done {
+			blueGreenStatus.CurrentStepState = v1beta1.CanaryStepStatePaused
+			klog.Infof("rollout(%s/%s) step(%d) state from(%s) -> to(%s)", c.Rollout.Namespace, c.Rollout.Name,
+				blueGreenStatus.CurrentStepIndex, v1beta1.CanaryStepStateMetricsAnalysis, blueGreenStatus.CurrentStepState)
+		}
+
+	case v1beta1.CanaryStepStatePaused:
+		klog.Infof("rollout(%s/%s) run bluegreen strategy, and state(%s)", c.Rollout.Namespace, c.Rollout.Name, v1beta1.CanaryStepStatePaused)
+		done, err := m.doCanaryPaused(c)
+		if err != nil {
+			return err
+		} else if done {
+			blueGreenStatus.LastUpdateTime = &metav1.Time{Time: time.Now()}
+			blueGreenStatus.CurrentStepState = v1beta1.CanaryStepStateReady
+			klog.Infof("rollout(%s/%s) step(%d) state from(%s) -> to(%s)", c.Rollout.Namespace, c.Rollout.Name,
+				blueGreenStatus.CurrentStepIndex, v1beta1.CanaryStepStatePaused, blueGreenStatus.CurrentStepState)
+		}
+
+	case v1beta1.CanaryStepStateReady:
+		klog.Infof("rollout(%s/%s) run bluegreen strategy, and state(%s)", c.Rollout.Namespace, c.Rollout.Name, v1beta1.CanaryStepStateReady)
+		// run next step
+		if len(c.Rollout.Spec.Strategy.BlueGreen.Steps) > int(blueGreenStatus.CurrentStepIndex) {
+			blueGreenStatus.LastUpdateTime = &metav1.Time{Time: time.Now()}
+			blueGreenStatus.CurrentStepIndex++
+			blueGreenStatus.NextStepIndex = util.NextBatchIndex(c.Rollout, blueGreenStatus.CurrentStepIndex)
+			blueGreenStatus.CurrentStepState = v1beta1.CanaryStepStateInit
+			klog.Infof("rollout(%s/%s) bluegreen step from(%d) -> to(%d)", c.Rollout.Namespace, c.Rollout.Name, blueGreenStatus.CurrentStepIndex-1, blueGreenStatus.CurrentStepIndex)
+		} else {
+			klog.Infof("rollout(%s/%s) bluegreen run all steps, and completed", c.Rollout.Namespace, c.Rollout.Name)
+			blueGreenStatus.LastUpdateTime = &metav1.Time{Time: time.Now()}
+			blueGreenStatus.CurrentStepState = v1beta1.CanaryStepStateCompleted
```

Comment:
can we fallthough to the case CanaryStepStateCompleted to avoid unnecessary reenqueue

Reply:
yes